rust/worker/src/segment/types.rs

```diff
@@ -136,131 +140,274 @@ pub struct MaterializedLogRecord<'referred_data> {
+
+        let final_embedding_at_log_index = if log_record.record.embedding.is_some() {
+            Some(log_index)
+        } else {
+            return Err(LogMaterializerError::EmbeddingMaterialization);
+        };
+
+        let merged_metadata;
+        let deleted_metadata;
+        match &log_record.record.metadata {
+            Some(metadata) => match materialize_update_metadata(metadata) {
+                Ok(m) => {
+                    merged_metadata = Some(m.0);
+                    deleted_metadata = Some(m.1);
+                }
+                Err(e) => {
+                    return Err(LogMaterializerError::MetadataMaterialization(e));
+                }
             },
-        }
+            None => {
+                merged_metadata = None;
+                deleted_metadata = None;
+            }
+        };
+
+        Ok(Self {
+            offset_id_exists_in_segment: false,
+            offset_id,
+            user_id_at_log_index: Some(log_index),
+            final_operation: MaterializedLogOperation::AddNew,
+            metadata_to_be_merged: merged_metadata,
+            metadata_to_be_deleted: deleted_metadata,
+            final_document_at_log_index,
+            final_embedding_at_log_index,
+        })
     }
+}
 
-    pub(crate) fn merged_document_ref(&self) -> Option<&str> {
-        if self.final_operation == MaterializedLogOperation::OverwriteExisting
-            || self.final_operation == MaterializedLogOperation::AddNew
-        {
-            return match self.final_document {
-                Some(doc) => Some(doc),
-                None => None,
-            };
-        }
-        match self.final_document {
-            Some(doc) => Some(doc),
-            None => match self.data_record.as_ref() {
-                Some(data_record) => match data_record.document {
-                    Some(doc) => Some(doc),
-                    None => None,
-                },
+/// Obtained from a `MaterializeLogsResult`. Provides a borrowed view of a single materialized log record.
+/// You will probably need to call `.hydrate()` on this struct for most use cases, although you can view the offset ID and operation without hydrating.
+pub struct BorrowedMaterializedLogRecord<'log_data> {
+    materialized_log_record: &'log_data MaterializedLogRecord,
+    logs: &'log_data Chunk<LogRecord>,
+}
+
+impl<'log_data> BorrowedMaterializedLogRecord<'log_data> {
+    pub fn get_offset_id(&self) -> u32 {
+        self.materialized_log_record.offset_id
+    }
+
+    pub fn get_operation(&self) -> MaterializedLogOperation {
+        self.materialized_log_record.final_operation
+    }
+
+    /// Reads any record segment data that this log record may reference and returns a hydrated version of this record.
+    /// The record segment reader passed here **must be over the same set of blockfiles** as the reader that was originally passed to `materialize_logs()`. If the two readers are different, the behavior is undefined.
```

Comment:
does not have to be incorporated in this PR but just some food for thought - can't we ship the reader together with the mat records? Readers are trivially send iiuc

Reply:
Yes, readers are Sendable, but have a lifetime, so that would require adding a lifetime to `MaterializeLogsResult`. Easy to do but potentially makes upstream types a little messier.